package/MDAnalysis/analysis/base.py

```diff
@@ -125,14 +125,105 @@
 import inspect
 import logging
 import itertools
+import warnings
 
 import numpy as np
 from MDAnalysis import coordinates
 from MDAnalysis.core.groups import AtomGroup
 from MDAnalysis.lib.log import ProgressBar
+from typing import Callable, Iterable
+from importlib import import_module
+from functools import partial
 
 logger = logging.getLogger(__name__)
 
+class ParallelExecutor(object):
+    """Class that executes parallel computations for AnalysisBase (but potentially for other purposes)
+    using one of the available backends"""
+
+    @classmethod
+    @property
+    def possible_backends(cls):
+        return ("local", "dask", "multiprocessing", "dask.distributed")
+
+    def __init__(self, backend: str, n_workers: int, client: object = None):
+        # validate 'client' argument
+        if client is not None and backend is not None:
+            raise ValueError(
+                (f"If 'client' is not None, backend is deduced and must be None, but here "
+                f'{backend=}, {client=}')
+            )
+        if client is not None:
+            # assuming it's a dask distributed client
+            from dask.distributed import Client as DistributedClient
+            if not isinstance(client, DistributedClient):
+                raise ValueError(f"client should be dask.distributed.Client instance, got {type(client)}")
+        self.client = client
+
+        # validate 'backend' argument
+        if backend is not None:
+            if backend not in ParallelExecutor.possible_backends:
+                raise ValueError(
+                    f"Invalid {backend=}: should be one of {ParallelExecutor.possible_backends}"
+                )
+            if backend == "local":
+                if n_workers > 1:
+                    raise ValueError("Can be only 1 worker with backend='local', got {n_workers=")
+            else:
+                try:
+                    import_module(backend)
+                except ImportError:
+                    raise ValueError(
+                        "Please install {backend} module "
+                        "with 'python3 -m pip install {backend} "
+                        "to run this")
+        self.backend = backend
+
+        # validate 'n_workers' argument
+        if n_workers and n_workers < 0:
+            raise ValueError(f"n_workers should be non-negative, got {n_workers=}")
+        self.n_workers = n_workers
+
+    def apply(self, func: Callable, computations: list) -> list:
+        if self.backend == "local":
```

Comment:
I don't really have a super constructive comment there. This is _a long_ if/else branch, it's only going to grow and become more painful. Anyone got any ideas on how to avoid this? A dictionary maybe?

Reply:
I'm unsettled too by this long list of `elif`s. A dictionary might be a good solution, but I can't think of a better way right now.

Reply:
The question in going to ask here, and I'm happy to have a longer offline conversation about it.

Do we absolutely *need* all these backends & options? I realise it's fun and offers flexibility, but from my perspective what we're doing here is working ourselves into a future maintenance nightmare.

Reply:
Sorry, I don't fully understand what's bad with these elifs, especially given that some of them require a different running scenario, like `dask.distributed`.

From what I understand @IAlibay is suggesting, it could look like this:

```python
def available_schedulers(self):
    return {'multiprocessing':self._compute_with_multiprocessing, 'dask':self._compute_with_dask, ...}
    
 def apply(self, func, computations):
     callable = self.available_schedulers.get(self.backend)
     if callable is None:
         raise ValueError
     return callable(func, computations)
```

but I believe it's slightly harder to work with, especially given that the dictionary size isn't going to increase significantly any time soon, and having 5 options in a flat if-elif-else is more explicit than hiding them in a dictionary values.

Reply:
>Do we absolutely need all these backends & options?

Well, this class isn't really supposed to be used by users right off the bat. For the user, there are essentially 3 backend mechanisms:
 - `multiprocessing` -- which has an advantage of being built-in
 - `dask` -- an extra dependency, but more flexible serialization mechanism than `multiprocessing`
 - `dask.distributed` -- for those 1% who a) are planning to analyze really huge trajectories b) know how to configure a `dask.distributed` cluster
 
I assume 90% of people will try to use (and be happy with) `multiprocessing`, then if it's not available for their task, switch to `dask`, and if it's still slow -- run distributed version.

Reply:
Re: if/else - it's a code smell. Large if/else branches (7 here) hints to a non-optimal design choice. For example, what happens if you need to keep adding options? The current design is such that the tree grows and then it becomes very hard to maintain / make work.

It doesn't necessarily mean that it's the wrong code decision, but it's always good to query these things and check if an alternative design might be more appropriate (not necessarily a dictionary, but any kind of early validation etc..).

Reply:
> Well, this class isn't really supposed to be used by users right off the bat

My comment here isn't about what users do, but rather what the maintenance burden will look like for us. The wider the number of backends the wider the maintenance costs. If they are all necessary that's an acceptable cost, but without any prior information here it's something I have to ask.